cpp/src/utilities/column_utils.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "column_utils.hpp"
+#include <utilities/error_utils.hpp>
+
+// This should go up into more general utility code...
+static constexpr inline bool logical_xor(bool x, bool y) { return (x and (not y)) or ((not x) and y) ; }
+
+namespace cudf {
+
+/**
+ * @brief Ensures a gdf_column is valid, i.e. that its fields are consistent
+ * with each other, and logical in themselves, in representing a proper column.
+ */
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "No column data despite there supposedly being rows.");
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type.");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column supposedly has null values, but no place to expose its record");
+    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column claims to have more null elements than elements overall");
+}
+
+void validate(const gdf_column* column_ptr)
+{
+    CUDF_EXPECTS ( (column_ptr != nullptr) , "Null gdf_column pointer");
+    validate(*column_ptr);
+}
+
+namespace detail {
+
+bool extra_type_info_is_compatible(
+    const gdf_dtype& common_dtype,
+    const gdf_dtype_extra_info& lhs_extra_type_info,
+    const gdf_dtype_extra_info& rhs_extra_type_info) noexcept
+{
+    switch(common_dtype) {
+    case GDF_CATEGORY:  return lhs_extra_type_info.category == rhs_extra_type_info.category;
+    case GDF_TIMESTAMP: return lhs_extra_type_info.time_unit == rhs_extra_type_info.time_unit;
+    default:            return true;
+    }
+}
+
+} // namespace detail
+
+
+bool have_same_type(const gdf_column& validated_column_1, const gdf_column& validated_column_2, bool ignore_extra_type_info) noexcept
+{
+    if (validated_column_1.dtype != validated_column_2.dtype) { return false; }
+    if (logical_xor(is_nullable(validated_column_1), is_nullable(validated_column_2))) { return false; }
```

Comment:
Are columns really different types if one has a validity mask and the other does not? What if one is an input, and another is a transformation of that input that results in it having validity? These columns are now not the same type?

Reply:
Of course they aren't. In type-theoretic terms, one has type τ and the other has type τ ⊍ Unit (τ disjoint-union with a unit type). Not the same thing at all.

But types don't have to be the same for every single operator - as they are convertible; and we can have them implicitly conversion if we want. So the fact that the two columns in your example are not the same type should not be a problem.

Reply:
I'm arguing that this line (the logical_xor) should not be part of `have_same_type`.

Reply:
> So the fact that the two columns in your example are not the same type should not be a problem.

Then why is it part of the test?

Reply:
@harrism : Because, like I said, the two columns in your example are of different types. This is not a test of whether one type's domain is a strict superset of the other's / injective convertibility etc.

Reply:
I wrote:
> Are columns really different types if one has a validity mask and the other does not? 
To which you answered:
> Of course they aren't.
Then you wrote:
> the two columns in your example are of different types.

Hmmmm :) I don't need the type theory. I'm trying to determine if this function is actually useful as-is. I can imagine wanting to compare the dtype/dtypeinfo of two columns, ignoring whether they are nullable. Since you didn't put this function into use anywhere, it's hard to know whether it's useful as-is.

I don't like adding functions that aren't used.

Reply:
@harrism : Some API functions expect different argument columns to, well, have the same type. Right now it's a lot of functions, since we use out-parameters; but even if that decreases, there will still be a few.

I only wrote utility functions I was using myself in API function implementations. While its true that both my PRs for those (multisearch, apply boolean mask) have not yet been merged, these utilities are not specific to just those two.